### PR TITLE
Fix last frame issue

### PR DESCRIPTION
**Pull request by sparshg**
## Motivation

The last frame for animations with updaters was not getting rendered in some cases. For example, if there is no animation after updaters are cleared
## Overview / Explanation for Changes

Digging down the previous commits, this was working fine in 3b1b version, so after searching I found that the if statement inside [finish_animations()](https://github.com/ManimCommunity/manim/commit/8b2e34db82a5d26f96ef49db22472896ecd2324e) was removed which is responsible for this issue. I re-added that with some changes to work correctly with current version.

## Example
<details><summary>Code</summary>

```py
from manim import *
from numpy import math

class Vectors(Scene):
    def construct(self):
        v = 10
        r = 0.5
        ang_vel = v/r
        circle = Circle(radius=r).shift(r*UP)
        dot = Dot(color=GREEN).shift(2*r*UP)
        self.add(NumberPlane(), circle, dot)
        time = ValueTracker(0)
         
        def dot_updater(m):
            m.move_to(r * np.array([math.sin(ang_vel * time.get_value()), math.cos(ang_vel * time.get_value()), 0]) + circle.get_center())
            # time.increment_value(dt)
        circle.add_updater(lambda m: m.move_to(RIGHT * v * time.get_value() + r*UP))
        dot.add_updater(dot_updater)
        self.add(circle, dot)
        self.play(ApplyMethod(time.set_value, TAU/ang_vel), rate_func=linear)
        dot.clear_updaters()
        circle.clear_updaters()
        print(dot.get_center())
        print(circle.get_center())
        self.wait()
```
</details>

### Before
https://user-images.githubusercontent.com/43041139/106367547-6e68a600-6369-11eb-8aeb-4098087e0208.mp4

### After
https://user-images.githubusercontent.com/43041139/106367555-7c1e2b80-6369-11eb-97d4-c0d9be243606.mp4

## Oneline Summary of Changes

```
- Last frame not rendered in some cases bug fixed (:pr:`983`)
```

## 

## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)

## Reviewer Checklist
- [ ] Newly added functions/classes are either private or have a docstring
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [ ] The oneline summary has been included [in the wiki](https://github.com/ManimCommunity/manim/wiki/Changelog-for-next-release)

### CODE REVIEW

**Comment by leotrs** [comment]
We _really really really_ should add a test for this.

**Comment by sparshg** [comment]
I have no idea how to do that

**Comment by leotrs** [comment]
There's some instructions in the documentation and/or wiki. Cc @ManimCommunity/tests

File: manim/scene/scene.py [hunk 1]
```diff
@@ -849,6 +849,8 @@ def play_internal(self, skip_rendering=False):
         for animation in self.animations:
             animation.finish()
             animation.clean_up_from_scene(self)
+        if not self.renderer.skip_animations:
+            self.update_mobjects(0)
```
Comment (huguesdevimeux) [inline manim/scene/scene.py:853]: **Question** : why 0? why updating at the last t?

Comment (sparshg) [reply manim/scene/scene.py:853]: I don't really know much details (because I found this by going back into commits history), but after an animation is finished we need to update the mobjects to avoid this bug. To update the mobjects now, update_mobjects(0) calls mobjects.update(0), which will set dt=0. (btw default value for dt is 0 if nothing is mentioned when calling mobjects.update())

Comment (huguesdevimeux) [reply manim/scene/scene.py:853]: I see your point, and I don't think it does have any importance but I think that it'd make more sense to update with the last t, no? 
Did you try?

Comment (sparshg) [reply manim/scene/scene.py:853]: I thought the same but didn't try coz of lack of time, got exams this week

Comment (huguesdevimeux) [reply manim/scene/scene.py:853]: No problem! 

good luck !

**Comment by behackl** [comment]
I've tested this locally now -- and I have to admit, it is very weird. The problem does not seem to occur within the testing framework; the image saved there from the doctest is already correct:
test
(and the test always produces the same image, independent of whether the change from this PR is included or not).

However, when rendering the test scene directly (with `-qm -s`), I get
LastFrameWhenClearedTest
(with the change from this PR included), and
LastFrameWhenClearedTest
without the change.

Weird.

**Comment by sparshg** [comment]
It might be related to #974 because in some situations, cache fixes the animations

**Comment by sparshg** [comment]
Yep, when you run the code for the test again (using cache), it gets fixed automatically...

**Comment by sparshg** [comment]
> The problem does not seem to occur within the testing framework; the image saved there from the doctest is already correct:
The problem IS in the testing framework. The cause of this testing issue, is the same the the cause of the cache issue, which is #550
When we skip the frames to get the last frame image, the updaters also skip the mid values and directly skip to final value, as described in #550. This is the same reason for the cache related problems with updaters.
In short, updaters tests are already broken, if you try adding [PointWithTrace](https://docs.manim.community/en/v0.2.0/examples.html#pointwithtrace) example as test, it will produce the wrong image to test. It is more of a coincidence that the existing updater tests work correctly because the mid-values do not matter in those and are very simple.
Now that is the exact reason why you see that dot inside the square... yes, the test just sets the updater value to the final value and gets the frame, which is obviously not the right way to produce the last frame... BUT, `-qm -s` produces the right result because if you notice, the animations are rendered _everytime_ **even with cache**

**TLDR:** Testing system broken, produces last frame the wrong way, should have been like -qm -s. Same reason for caching system bug #550

File: manim/scene/scene.py (hunk 1 above)
Comment (huguesdevimeux) [reply manim/scene/scene.py:853]: So, did you try it?

Comment (sparshg) [reply manim/scene/scene.py:853]: isn't the mobjects.update(0) already only running when an animation ends?

**Review by huguesdevimeux — APPROVED** [review]

**Comment by behackl** [comment]
@sparshg, do you think this can still be merged for now?

**Review by behackl — APPROVED** [review]

**Comment by sparshg** [comment]
yep, as I said the issue is in testing system for now going through intermediate frames, thats not related with this PR. The test included with this PR is pointless for the time the testing system is broken, but again thats another issue...
